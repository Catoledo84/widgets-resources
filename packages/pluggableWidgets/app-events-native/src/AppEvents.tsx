import NetInfo, { NetInfoSubscription, NetInfoState } from "@react-native-community/netinfo";
import { Component } from "react";
import { AppState, AppStateStatus } from "react-native";

import { AppEventsProps } from "../typings/AppEventsProps";
import { executeAction } from "@mendix/piw-utils-internal";

export type Props = AppEventsProps<undefined>;

export class AppEvents extends Component<Props> {
    private readonly onAppStateChangeHandler = this.onAppStateChange.bind(this);
    private readonly onConnectionChangeHandler = this.onConnectionChange.bind(this);

    private appState = AppState.currentState;
    private lastOnResume = Date.now();

    private isConnected?: boolean;
    private lastOnOnline = 0;
    private lastOnOffline = 0;
    private onLoadTriggered = false;
    private timeoutHandle?: any;
    private unsubscribeNetworkEventListener?: NetInfoSubscription;
    private scheduleTimer?: typeof setTimeout | typeof setInterval;
    private clearTimer?: typeof clearTimeout | typeof clearInterval;

    componentDidMount(): void {
        this.scheduleTimer = this.props.timerType === "once" ? setTimeout : setInterval;
        this.clearTimer = this.props.timerType === "once" ? clearTimeout : clearInterval;

        AppState.addEventListener("change", this.onAppStateChangeHandler);

        this.scheduleOnTimeOutEvent();

        if (!this.onLoadTriggered && this.props.onLoadAction?.canExecute) {
            this.onLoadTriggered = true;
            executeAction(this.props.onLoadAction);
        }

        // Not awaited on purposes. The NetInfo request might take a long time to complete.
        // Aside as this is an async event to be called; there's no particular value
        // in blocking until done. If noteworthy it will fire an event.
        this.subscribeToNetworkStatus();
    }

    componentWillUnmount(): void {
        if (this.props.onUnloadAction && this.props.onUnloadAction.canExecute) {
            executeAction(this.props.onUnloadAction);
        }

        if (this.props.onResumeAction) {
            AppState.removeEventListener("change", this.onAppStateChangeHandler);
        }

        if (this.unsubscribeNetworkEventListener) {
            this.unsubscribeNetworkEventListener();
        }

        if (this.props.onTimeoutAction && this.timeoutHandle != null) {
            this.clearTimer?.(this.timeoutHandle);
        }
    }

    render(): null {
        return null;
    }

    private onAppStateChange(nextAppState: AppStateStatus): void {
        if (this.appState === nextAppState) {
            return;
        }
        if (nextAppState === "active") {
            this.scheduleOnTimeOutEvent();
            // The app was paused in a particular state.
            // Skipping updating isConnected state and expect the network listener
            // to trigger if the state actually changed since the pause event.
            this.subscribeToNetworkStatus(false);

            if (isPastTimeout(this.lastOnResume, this.props.onResumeTimeout)) {
                executeAction(this.props.onResumeAction);
                this.lastOnResume = Date.now();
            }
        }

        if (nextAppState === "background") {
            this.unsubscribeFromNetworkStatus();
            this.unscheduleOnTimeoutEvent();
        }

        this.appState = nextAppState;
    }

    private onConnectionChange(nextState: NetInfoState): void {
        if (this.isConnected === nextState.isConnected) {
            return;
        }

        if (nextState.isConnected && isPastTimeout(this.lastOnOnline, this.props.onOnlineTimeout)) {
            executeAction(this.props.onOnlineAction);
            this.lastOnOnline = Date.now();
        }

        if (!nextState.isConnected && isPastTimeout(this.lastOnOffline, this.props.onOfflineTimeout)) {
            executeAction(this.props.onOfflineAction);
            this.lastOnOffline = Date.now();
        }

        this.isConnected = nextState.isConnected;
    }

    private async subscribeToNetworkStatus(updateIsConnectedState = true): Promise<void> {
        this.unsubscribeFromNetworkStatus();
        if (this.props.onOnlineAction || this.props.onOfflineAction) {
            if (updateIsConnectedState) {
                this.isConnected = (await NetInfo.fetch()).isConnected;
            }
            this.unsubscribeNetworkEventListener = NetInfo.addEventListener(this.onConnectionChangeHandler);
        }
    }

    private unsubscribeFromNetworkStatus(): void {
        this.unsubscribeNetworkEventListener?.();
    }

    private scheduleOnTimeOutEvent(): void {
        this.unscheduleOnTimeoutEvent();
        if (this.props.onTimeoutAction) {
            this.timeoutHandle = this.scheduleTimer?.(
                () => executeAction(this.props.onTimeoutAction),
                this.props.delayTime * 1000
            );
        }
    }

    private unscheduleOnTimeoutEvent(): void {
        if (this.timeoutHandle) {
            this.clearTimer?.(this.timeoutHandle);
        }
        this.timeoutHandle = null;
    }
}

function isPastTimeout(last: number, timeoutSeconds: number): boolean {
    return Date.now() - last >= timeoutSeconds * 1000;
}
