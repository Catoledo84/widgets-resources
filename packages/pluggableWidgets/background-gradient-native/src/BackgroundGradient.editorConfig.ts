import { StructurePreviewProps, Problem } from "@mendix/piw-utils-internal";
import { BackgroundGradientPreviewProps } from "../typings/BackgroundGradientProps";
import iconLight from "./assets/backgroundGradient_light.svg";
import iconDark from "./assets/backgroundGradient_dark.svg";

export const getPreview = (values: BackgroundGradientPreviewProps, isDarkMode: boolean): StructurePreviewProps => ({
    type: "Container",
    borders: true,
    children: [
        {
            type: "RowLayout",
            columnSize: "grow",
            borders: true,
            padding: 0,
            backgroundColor: isDarkMode ? "#454545" : "#F5F5F5",
            children: [
                {
                    type: "Container",
                    children: [
                        {
                            type: "RowLayout",
                            columnSize: "grow",
                            children: [
                                {
                                    type: "Container",
                                    padding: 5,
                                    children: [
                                        {
                                            type: "Image",
                                            document: decodeURIComponent(
                                                (isDarkMode ? iconDark : iconLight).replace("data:image/svg+xml,", "")
                                            ),
                                            width: 16,
                                            height: 16
                                        }
                                    ]
                                },
                                {
                                    type: "Text",
                                    fontSize: 12,
                                    fontColor: "#6B707B",
                                    content: "Gradient Background"
                                }
                            ]
                        }
                    ]
                },
                { type: "Container", grow: 2 }
            ]
        },
        {
            type: "DropZone",
            property: values.content as object,
            placeholder: "Configure your background gradient ( Content )"
        }
    ]
});

export function check(values: BackgroundGradientPreviewProps): Problem[] {
    const errors: Problem[] = [];
    const { opacity, angle, colorList } = values;
    if (opacity) {
        if (opacity > 1) {
            errors.push({
                property: "opacity",
                message: "opacity can't be more than 1"
            });
        } else if (opacity < 0) {
            errors.push({
                property: "opacity",
                message: "opacity can't be less than 0"
            });
        }
    }
    if (angle) {
        if (angle > 360) {
            errors.push({
                property: "angle",
                message: "angle can't be more than 360"
            });
        } else if (angle < 0) {
            errors.push({
                property: "angle",
                message: "angle can't be less than 0"
            });
        }
    }

    if (colorList.some(item => item.offset! > 1 || item.offset! < 0)) {
        errors.push({
            property: "colorList",
            message: "color offset should be between 0 and 1"
        });
    }

    // if (!values.advancedMode) {
    //     if (!values.basicItems.length) {
    //         errors.push({
    //             property: "basicItems",
    //             message: "For the popup menu to be visible, you need to add at least one item to it."
    //         });
    //     }
    //     values.basicItems.forEach((item, index) => {
    //         if (item.itemType === "item" && !item.caption) {
    //             errors.push({
    //                 property: `basicItems/${index + 1}/caption`,
    //                 message: "The 'Caption' property is required."
    //             });
    //         }
    //     });
    // } else {
    //     if (!values.customItems.length) {
    //         errors.push({
    //             property: "customItems",
    //             message: "For the popup menu to be visible, you need to add at least one item to it."
    //         });
    //     }
    // }
    return errors;
}
