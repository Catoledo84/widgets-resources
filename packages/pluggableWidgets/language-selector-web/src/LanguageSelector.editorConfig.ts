import { Properties, StructurePreviewProps } from "@mendix/piw-utils-internal";
import { LanguageSelectorPreviewProps } from "typings/LanguageSelectorProps";
import ArrowDark from "./assets/arrow_dark.svg";
import ArrowLight from "./assets/arrow_light.svg";

export function getProperties(__values: LanguageSelectorPreviewProps, defaultValues: Properties): Properties {
    return defaultValues;
}

export function getPreview(__values: LanguageSelectorPreviewProps, isDarkMode: boolean): StructurePreviewProps | null {
    return {
        type: "RowLayout",
        columnSize: "grow", // dynamic column sizes
        children: [
            {
                type: "Text",
                content: "Current language",
                fontColor: isDarkMode ? "#DEDEDE" : "#899499"
            },
            { type: "Container", grow: 0.001 }, // small space between items
            {
                type: "Container",
                padding: 6,
                children: [
                    {
                        type: "Image",
                        document: decodeURIComponent(
                            (isDarkMode ? ArrowLight : ArrowDark).replace("data:image/svg+xml,", "")
                        ),
                        height: 8
                    }
                ]
            }, // fills space on right

            { type: "Container" } // fills space on right
        ]
    };
}
