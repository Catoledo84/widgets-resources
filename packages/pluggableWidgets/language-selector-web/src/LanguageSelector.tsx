import { ReactNode, createElement, useEffect, useState, useCallback } from "react";
import { LanguageSelectorContainerProps } from "typings/LanguageSelectorProps";
import { PopupMenu } from "./Popup/PopupMenu";

export type AttributeView = {
    value: string;
    key: string;
};
export type LanguageAttribute = {
    Code: AttributeView;
    Description: AttributeView;
};
export type LanguageItem = {
    _guid: string;
    jsonData: {
        attributes: LanguageAttribute;
    };
};
export default function LanguageSelector(props: LanguageSelectorContainerProps): ReactNode {
    const [language, setLanguage] = useState<LanguageItem | null>(null);
    const [languageList, setLanguageList] = useState<LanguageItem[]>([]);
    useEffect(() => {
        const getLanguages = (): void => {
            window.mx.data.get({
                xpath: "//System.Language",
                callback: (objs: LanguageItem[]) => {
                    if (objs.length) {
                        setLanguageList(objs);
                        const currentLanguage: LanguageItem | null =
                            objs.find(
                                (lang: LanguageItem) =>
                                    lang._guid ===
                                    window.mx.session.getUserObject()?.jsonData.attributes["System.User_Language"].value
                            ) || null;
                        console.log(currentLanguage);
                        setLanguage(currentLanguage);
                    }
                },
                error: e => console.log("errr", e)
            });
        };
        getLanguages();
    }, [window.mx]);

    const selectLanguage = useCallback((item: LanguageItem) => {
        setLanguage(item);
        window.mx.session.getUserObject()?.addReference("System.User_Language", "" + item._guid);
        window.mx.data.commit({
            mxobj: window.mx.session.getUserObject(),
            callback() {
                // window.mx.session.sessionStore.remove().then(() => window.mx.reloadWithState());
                window.mx.session.clearCachedSessionData().then(() => window.mx.reload());
            },
            error: e => {
                console.error("Could not commit object:", e);
            }
        });
    }, []);
    return (
        <div>
            {mx.session.getUserObject()?.addReference ? (
                <PopupMenu
                    position={props.position}
                    currentLanguage={language}
                    languageList={languageList}
                    onSelect={selectLanguage}
                    trigger={props.trigger}
                ></PopupMenu>
            ) : null}
        </div>
    );
}
