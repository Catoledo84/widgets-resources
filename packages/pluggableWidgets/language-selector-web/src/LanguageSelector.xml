<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.languageSelector.LanguageSelector" needsEntityContext="true" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Language selector</name>
    <description />
    <studioProCategory>Display</studioProCategory>
    <studioCategory>Display</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/widgets/languageSelector</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Languages">
                <property key="languageOptions" type="datasource" isList="true" required="true">
                    <caption>Data source</caption>
                    <description />
                </property>
                <property key="languageCaption" type="textTemplate" dataSource="languageOptions" required="true">
                    <caption>Language caption</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="General">
                <property key="position" type="enumeration" defaultValue="top">
                    <caption>Menu position</caption>
                    <description>How to position the menu in relation to the trigger element - at the top, to the left, at the bottom or to the right.</description>
                    <enumerationValues>
                        <enumerationValue key="left">Left</enumerationValue>
                        <enumerationValue key="right">Right</enumerationValue>
                        <enumerationValue key="top">Top</enumerationValue>
                        <enumerationValue key="bottom">Bottom</enumerationValue>
                    </enumerationValues>
                </property>

                <property key="trigger" type="enumeration" defaultValue="click">
                    <caption>Open on</caption>
                    <description>How the menu is triggered - click, hover.</description>
                    <enumerationValues>
                        <enumerationValue key="click">Click</enumerationValue>
                        <enumerationValue key="hover">Hover</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="hideForSingle" type="boolean" defaultValue="true">
                    <caption>Hide for single language</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
