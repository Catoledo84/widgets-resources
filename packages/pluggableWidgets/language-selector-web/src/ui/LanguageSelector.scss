.popupmenu-trigger-alignement {
    display: flex;
    align-items: center;
}
.current-language-text {
    margin-right: 6px;
}
