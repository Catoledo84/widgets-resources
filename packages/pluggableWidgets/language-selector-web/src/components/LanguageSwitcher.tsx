import classNames from "classnames";
import { createElement, ReactElement, useEffect, useRef, useState, useCallback } from "react";
import {
    isBehindElement,
    isBehindRandomElement,
    isElementPartiallyOffScreen,
    isElementVisibleByUser,
    moveAbsoluteElementOnScreen,
    unBlockAbsoluteElementBottom,
    unBlockAbsoluteElementLeft,
    unBlockAbsoluteElementRight,
    unBlockAbsoluteElementTop,
    handleOnClickOutsideElement
} from "@mendix/piw-utils-internal";
import { PositionEnum, TriggerEnum } from "../../typings/LanguageSelectorProps";
import { LanguageItem } from "../LanguageSelector";

export interface LanguageSwitcherProps {
    preview: boolean;
    currentLanguage: LanguageItem | undefined;
    languageList: LanguageItem[];
    position: PositionEnum;
    onSelect: (lang: LanguageItem) => void;
    trigger: TriggerEnum;
}
export const LanguageSwitcher = (props: LanguageSwitcherProps): ReactElement => {
    const preview = !!props.preview;
    const { languageList, trigger } = props;
    const [visibility, setVisibility] = useState(false);
    const ref = useRef<HTMLDivElement>(null);
    if (!preview) {
        handleOnClickOutsideElement(ref, () => setVisibility(false));
    }
    const menuOptions = (): ReactElement[] => {
        return languageList.map(item => {
            return (
                <div key={item._guid} className={"popupmenu-basic-item"} onClick={() => props.onSelect(item)}>
                    {item.value}
                </div>
            );
        });
    };

    const onClickHandle = useCallback(
        (e): void => {
            e.preventDefault();
            e.stopPropagation();
            setVisibility(prev => !prev);
        },
        [setVisibility]
    );

    const onHover =
        trigger === "hover"
            ? {
                  onMouseEnter: onClickHandle,
                  onMouseLeave: onClickHandle
              }
            : {};

    const onClick = trigger === "click" ? { onClick: onClickHandle } : {};

    useEffect(() => {
        const element = ref.current?.querySelector(".popupmenu-menu") as HTMLDivElement | null;
        if (element) {
            element.style.display = visibility ? "flex" : "none";
            if (visibility) {
                correctPosition(element, props.position);
            }
        }
    }, [props.position, visibility]);

    return (
        <div ref={ref} className={classNames("popupmenu")} {...onHover}>
            <div className={"popupmenu-trigger"} {...onClick}>
                {props?.currentLanguage?.value || ""}
            </div>
            <div className={classNames("popupmenu-menu", `popupmenu-position-${props.position}`)}>{menuOptions()}</div>
        </div>
    );
};

function correctPosition(element: HTMLElement, position: PositionEnum): void {
    const dynamicDocument: Document = element.ownerDocument;
    const dynamicWindow = dynamicDocument.defaultView as Window;
    let boundingRect: DOMRect = element.getBoundingClientRect();
    const isOffScreen = isElementPartiallyOffScreen(dynamicWindow, boundingRect);
    if (isOffScreen) {
        moveAbsoluteElementOnScreen(dynamicWindow, element, boundingRect);
    }

    boundingRect = element.getBoundingClientRect();
    const blockingElement = isBehindRandomElement(dynamicDocument, element, boundingRect, 3, "popupmenu");
    if (blockingElement && isElementVisibleByUser(dynamicDocument, dynamicWindow, blockingElement)) {
        unBlockAbsoluteElement(element, boundingRect, blockingElement.getBoundingClientRect(), position);
    } else if (blockingElement) {
        let node = blockingElement;
        do {
            if (isBehindElement(element, node, 3) && isElementVisibleByUser(dynamicDocument, dynamicWindow, node)) {
                return unBlockAbsoluteElement(element, boundingRect, node.getBoundingClientRect(), position);
            } else if (node.parentElement) {
                node = node.parentElement as HTMLElement;
            } else {
                break;
            }
        } while (node.parentElement);
    }
}

function unBlockAbsoluteElement(
    element: HTMLElement,
    boundingRect: DOMRect,
    blockingElementRect: DOMRect,
    position: PositionEnum
): void {
    switch (position) {
        case "left":
            unBlockAbsoluteElementLeft(element, boundingRect, blockingElementRect);
            unBlockAbsoluteElementBottom(element, boundingRect, blockingElementRect);
            break;
        case "right":
            unBlockAbsoluteElementRight(element, boundingRect, blockingElementRect);
            unBlockAbsoluteElementBottom(element, boundingRect, blockingElementRect);
            break;
        case "top":
            unBlockAbsoluteElementTop(element, boundingRect, blockingElementRect);
            break;
        case "bottom":
            unBlockAbsoluteElementBottom(element, boundingRect, blockingElementRect);
            break;
    }
}
