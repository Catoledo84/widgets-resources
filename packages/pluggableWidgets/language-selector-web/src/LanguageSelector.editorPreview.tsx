import { createElement, ReactElement } from "react";
import { LanguageSelectorPreviewProps } from "typings/LanguageSelectorProps";
import { LanguageSwitcher } from "./components/LanguageSwitcher";

export const preview = (props: LanguageSelectorPreviewProps): ReactElement => {
    return (
        <LanguageSwitcher
            preview={false}
            currentLanguage={{ _guid: "1", value: "Current language" }}
            languageList={[{ _guid: "1", value: "Current language" }]}
            position={props.position}
            onSelect={undefined}
            trigger={props.trigger}
            className={""}
        ></LanguageSwitcher>
    );
};

export function getPreviewCss(): string {
    return require("./ui/LanguageSelector.scss");
}
