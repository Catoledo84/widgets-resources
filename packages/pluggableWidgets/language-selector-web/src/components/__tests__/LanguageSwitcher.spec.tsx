import { mount, render } from "enzyme";
import { createElement } from "react";
import { LanguageItem } from "src/LanguageSelector";
import { LanguageSwitcher } from "../LanguageSwitcher";

const props = {};
describe("Language switcher", () => {
    it("renders the structure with an image", () => {
        expect(
            render(
                <LanguageSwitcher
                    preview={false}
                    currentLanguage={undefined}
                    languageList={[]}
                    position={"left"}
                    onSelect={function (lang: LanguageItem): void {
                        throw new Error("Function not implemented.");
                    }}
                    trigger={"click"}
                    {...props}
                />
            )
        ).toMatchSnapshot();
    });
});
