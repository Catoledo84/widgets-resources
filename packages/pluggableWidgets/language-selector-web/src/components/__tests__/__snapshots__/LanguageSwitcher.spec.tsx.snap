// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language switcher renders renders the structure with language list and selected default language 1`] = `
<DocumentFragment>
  <div
    class="widget-language-selector popupmenu"
  >
    <div
      class="popupmenu-trigger popupmenu-trigger-alignement"
      style="display: flex; align-items: center;"
    >
      <span
        class="current-language-text"
      >
        En us
      </span>
      <span
        aria-hidden="true"
        class="glyphicon glyphicon-chevron-down"
      />
    </div>
    <div
      class="popupmenu-menu popupmenu-position-left"
    >
      <div
        class="popupmenu-basic-item"
      >
        En us
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Language switcher renders the structure with empty language list 1`] = `
<DocumentFragment>
  <div
    class="widget-language-selector popupmenu"
  >
    <div
      class="popupmenu-trigger popupmenu-trigger-alignement"
      style="display: flex; align-items: center;"
    >
      <span
        class="current-language-text"
      />
      <span
        aria-hidden="true"
        class="glyphicon glyphicon-chevron-down"
      />
    </div>
    <div
      class="popupmenu-menu popupmenu-position-left"
    />
  </div>
</DocumentFragment>
`;
